data-science-onramp/data-cleaning/clean_test.py

```diff
@@ -0,0 +1,158 @@
+import os
+import re
+import uuid
+
+from google.cloud import dataproc_v1 as dataproc
+from google.cloud import storage
+import pytest
+
+# Set global variables
+PROJECT = os.environ['GCLOUD_PROJECT']
+REGION = "us-east4"
+CLUSTER_NAME = f'clean-test-{uuid.uuid4()}'
+BUCKET_NAME = f'clean-test-code-{uuid.uuid4()}'
+DESTINATION_BLOB_NAME = "clean.py"
+JOB_FILE_NAME = f'gs://{BUCKET_NAME}/clean.py'
+JOB_DETAILS = {  # Job configuration
+    'placement': {
+        'cluster_name': CLUSTER_NAME
+    },
+    'pyspark_job': {
+        'main_python_file_uri': JOB_FILE_NAME,
+        'args': [
+            PROJECT,
+            BUCKET_NAME,
+            "--test",
+        ],
+        "jar_file_uris": [
+                "gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar"
+        ],
+    },
+}
+CLUSTER_DATA = {  # Create cluster configuration
+    'project_id': PROJECT,
+    'cluster_name': CLUSTER_NAME,
+    'config': {
+        'gce_cluster_config': {
+            'zone_uri': '',
+            "metadata": {
+                "PIP_PACKAGES": "google-cloud-storage"
+            },
+        },
+        'master_config': {
+            'num_instances': 1,
+            'machine_type_uri': 'n1-standard-8'
+        },
+        'worker_config': {
+            'num_instances': 6,
+            'machine_type_uri': 'n1-standard-8'
+        },
+        "initialization_actions": [
+            {
+                "executable_file": ("gs://dataproc-initialization-actions/"
+                                    "python/pip-install.sh"),
+            }
+        ],
+        "software_config": {
+            "image_version": "1.5.4-debian10",
+            "optional_components": [
+                "ANACONDA"
+            ],
+        }
+    }
+}
+
+@pytest.fixture(autouse=True)
+def setup_and_teardown_cluster():
+    # Create cluster using cluster client
+    cluster_client = dataproc.ClusterControllerClient(client_options={
+        'api_endpoint': f'{REGION}-dataproc.googleapis.com:443'
+    })
+    operation = cluster_client.create_cluster(PROJECT, REGION, CLUSTER_DATA)
+
+    # Wait for cluster to provision
+    operation.result()
+
+    yield
+
+    # Delete cluster
+    operation = cluster_client.delete_cluster(PROJECT, REGION, CLUSTER_NAME)
+    operation.result()
+
+@pytest.fixture(autouse=True)
+def setup_and_teardown_bucket():
+    # Create GCS Bucket
+    storage_client = storage.Client()
+    bucket = storage_client.create_bucket(BUCKET_NAME)
+
+    # Upload file
+    blob = bucket.blob(DESTINATION_BLOB_NAME)
+    blob.upload_from_filename("clean.py")
+
+    yield
+
+    # Delete GCS bucket
+    bucket = storage_client.get_bucket(BUCKET_NAME)
+    bucket.delete(force=True)
+
+def is_in_table(value, out):
+    return re.search(f"\\| *{value} *\\|", out)
+
+def get_blob_from_path(path):
```

Comment:
I feel like I asked this last time - sorry for forgetting - this is a dataproc specific thing, right? We can't pass this as an env var because Dataproc makes it along the way? @bradmiro

Reply:
Dataproc only gives you the URI for the output logs and GCS client libs don't support URI access. You need to parse out the bucket and blob (of which GCS has no native function to do this).